state-sync/storage-service/server/src/lib.rs

```diff
@@ -1163,28 +1246,62 @@ impl StorageReaderInterface for StorageReader {
-            return Err(Error::InvalidRequest(format!(
-                "Requested number of states is larger than the maximum! \
-             Requested: {:?}, maximum: {:?}.",
-                expected_num_states, max_state_chunk_size
-            )));
+        // Calculate the number of state values to fetch
+        let expected_num_state_values = inclusive_range_len(start_index, end_index)?;
+        let max_num_state_values = self.config.max_state_chunk_size;
+        let mut num_state_values_to_fetch = min(expected_num_state_values, max_num_state_values);
+
+        // Attempt to serve the request
+        while num_state_values_to_fetch >= 1 {
+            let state_value_chunk_with_proof = self
+                .storage
+                .get_state_value_chunk_with_proof(
+                    version,
+                    start_index as usize,
+                    num_state_values_to_fetch as usize,
+                )
+                .map_err(|error| Error::StorageErrorEncountered(error.to_string()))?;
+
+            let (overflow_frame, num_bytes) = check_overflow_network_frame(
+                &state_value_chunk_with_proof,
+                self.config.max_network_chunk_bytes,
+            )?;
+            if !overflow_frame {
+                return Ok(state_value_chunk_with_proof);
+            } else {
+                increment_network_frame_overflow(
+                    DataResponse::StateValueChunkWithProof(state_value_chunk_with_proof)
+                        .get_label(),
+                );
+                let new_num_state_values_to_fetch = num_state_values_to_fetch / 2;
+                debug!("The request for {:?} state values was too large (num bytes: {:?}). Retrying with {:?}.",
+                    num_state_values_to_fetch, num_bytes, new_num_state_values_to_fetch);
+                num_state_values_to_fetch = new_num_state_values_to_fetch; // Try again with half the amount of data
+            }
         }
 
-        let state_value_chunk_with_proof = self
-            .storage
-            .get_state_value_chunk_with_proof(
-                version,
-                start_index as usize,
-                expected_num_states as usize,
-            )
-            .map_err(|error| Error::StorageErrorEncountered(error.to_string()))?;
-        Ok(state_value_chunk_with_proof)
+        panic!(
+            "Unable to serve the get_state_value_chunk_with_proof request! Version: {:?}, \
+            start index: {:?}, end index: {:?}. The data cannot fit into a single network frame!",
+            version, start_index, end_index
+        )
     }
 }
 
+/// Serializes the given data and returns true iff the data will overflow
+/// the maximum network frame size. Also returns the number of serialized
+/// bytes for logging purposes.
+fn check_overflow_network_frame<T: ?Sized + Serialize>(
+    data: &T,
+    max_network_frame_bytes: u64,
+) -> Result<(bool, u64), Error> {
+    let num_serialized_bytes = bcs::to_bytes(&data)
```

Comment:
we should replace this once storage api supports maximum bytes directly

Reply:
Ack. Will wait for them to provide this 😄